AI/Nullkiller/Pathfinding/AINodeStorage.cpp

```diff
@@ -1203,7 +1203,10 @@ void AINodeStorage::calculateTownPortalTeleportations(std::vector<CGPathNode *>
 	std::vector<const ChainActor *> actorsVector(actorsOfInitial.begin(), actorsOfInitial.end());
 	tbb::concurrent_vector<CGPathNode *> output;
 
-	if(actorsVector.size() * initialNodes.size() > 1000)
+// TODO: re-enable after fixing thread races. See issue for details:
+// https://github.com/vcmi/vcmi/pull/4130
```

Comment:
Locking with a mutex after

auto bestNode = getBestInitialNodeForTownPortal(targetTown);

if(!bestNode)
	return std::nullopt;

should be enough

Reply:
understood, will take a look.

Reply:
Will try to fix it later